### DISCUSSION TITLE
v4.0.39 pre-release

### DISCUSSION THREAD

**Discussion by autoSteve** [BJReplay/ha-solcast-solar#93] [Announcements]
This release carries a fix, an enhancement, and a change.

The fix? Get rid of the `scipy` library.

Many users were reporting that it could not be downloaded on their HA deployments. I'd estimate 5% of v4.0.38 installs, so extrapolating for the > 3000 downloads of 'latest' v4.0.31, that would likely result in > 150 issue reports that have been avoided.

The enhancement? More nerd knobs.

In the integration `CONFIGURE` dialogues there is now a section to specify which sensor attributes are published, and this can remove the clutter of un-needed things, like a site breakdown for forecasts and `estimate50`/`estimate90`.

The change? Altering some weirdly named and described sensors. These are power in 30/60 minutes, and custom X hours sensor.

As always, both positive and negative feedback is most welcome, and this will contribute to getting the release ultimately published as latest. Report back, and don't be shy!

Feedback in the past has been gratefully received from @dtswk, @EddieSpaghettie, @ffm777, @gcoan, @hwb45, @iainfogg, @isorin, @jeffb1974, @kernow, @ladaan, @Markyt77, @nilogax, @Noodleyman, @ProphetOfDoom, @rholligan and @wimdebruyn, and continued thanks to them for their interest in beta releases (and alphas!) and for their contributions.

> [!NOTE]
> The BJReplay fork is **_still_** not published for searching in HACS, so continued use of a manual install in HACS is the go. As explained earlier we're going to keep any releases as pre-release on the fork until that happens, as any 'latest' version change would require modification of that HACS pull request.

## What's Changed
* Updates to sensor descriptions, and alter some sensor names by @isorin (Potentially breaking for UI/automations/etc. should these these sensors be in use. Power in 30/60 minutes, and custom X hours sensor.)
* Remove dependency on scipy library by @autoSteve
* Add granular configuration options for attributes by @autosteve

Full Changelog: https://github.com/BJReplay/ha-solcast-solar/compare/v4.0.38...v4.0.39

**Comment by dtswk** [comment]
Thanks @autoSteve bloody nice work.  Have you retained the grainular remaining today.?  ie updating per minute or has it gone back to half hour jumps?  Very happy to beta test.

**Reply by dtswk** [reply]
Just installed and remaining forecast today is in 5 minute increments.  Perfect, I played around with turning on and off the new configs no errors.  So far looks good, will report any issues.

**Comment by gcoan** [comment]
Reverted from the alpha to the bjreplay beta 4.0.39, no problems.  Running OK, sensor values look OK and no errors in the log.

Did the upgrade in a single HACS release and only 1 restart of HA:
- added bjreplay custom repository
- removed autosteve custom repository
- clicked on Solcast integration in HACS
- clicked download
- the download defaults to 0.31, turn the beta switch on, then select beta 4.0.38 but don't download.  Then select latest beta 4.0.39 and click download
- restart HA

There is something buggy with the HACS download dialogue, I've seen it before.  Selecting a version you don't want and then the version you do works around this weirdness

**Comment by EddieSpaghettie** [comment]
We have seem so much growth in this integration in the last 3/4 weeks. I have 0 complaints ans want to thank you all for all the awesome changes and new features. Keep it up!

**Comment by EddieSpaghettie** [comment]
One oddity. But not sure its related to the current version
 Sun just came up and it polled. Says forecast todag is 20.04 and remaining is 20.05. Not an issue for me. Just thought ill mention it.

**Reply by autoSteve** [reply]
I would expect the difference in the two results to be the other way round, like you point out, with remaining lower than total expected. This points out a flaw in the math.

Forecast remaining uses a cubic spline to calculate values for part-way into a leading half-hour, which if anything should result in a lower value, given the forward trend of "down" remaining, which would makes sense. The numbers in this case don't, so read on.

As @isorin rightly pointed out separately, "trapezoidal interpolation" utilising a **mid-point** as a basis for interpolated values should be used. It makes sense that this is behind what is going on in this instance, as the present code would use the leading value of the half hour and not treat it as a middle-of-half-hour, and this would account for the difference.

So thanks for pointing it out.

I'll think about and eventually right this. (But @isorin, do feel free to correct my ham-fisted math while I'm AFK playing in Tokyo...)

**Reply by isorin** [reply]
It could be just a simple rounding error.
It should correct itself if/when we implement more precise energy calculation.

**Comment by ProphetOfDoom** [comment]
V4.0.39 installed yesterday, and apart from deleting the now obsolete forecast_next_1_hour sensor, everything seems to be working AOK.

New config options showing just fine as well, though I haven't had the need or the time as yet to have a play with them. 

All looking VV Good!

![image](https://github.com/user-attachments/assets/ce764bcd-80ae-479e-bc1d-e9288076454e)

**Reply by gcoan** [reply]
> deleting the now obsolete forecast_next_1_hour 

Good point we should add this to the release instructions 

For me the sensor name is sensor.solcast_pv_forecast_next_1_hours (ie plural)

**Reply by isorin** [reply]
Is there a new sensor called "Forecast Next X Hours" to replace it?

**Reply by ProphetOfDoom** [reply]
@gcoan I wrote it down from memory, so probably approximated the name. I've deleted it now so can't easily go back and check. I'll try not to be so tardy next time!

@isorin Yup. Right there in the middle.

![image](https://github.com/user-attachments/assets/cc2cb11b-24b0-4736-9220-87b44bc56c0f)

**Reply by isorin** [reply]
Looks good, yeah, I think we should add to the release notes that some sensors are replaced with new ones and the old ones need to be deleted manually.

**Comment by autoSteve** [comment]
It's in the notes for the release now, but might need to make it into the changelog in the readme, which no one will read. They similarly won't read the notes in the release.

**Comment by gcoan** [comment]
I don't think this is a bug, because from what I can see the integration is doing the right thing, but more a matter of pointing out that peak forecast time for multi-site arrays can produce unexpected results:

![image](https://github.com/user-attachments/assets/fe028b78-4b4f-459f-8786-440294512e97)

When I first saw this I was surprised with the sensor peak time given the individual site peak times are so far apart. My two sites (arrays-ish) are East and West aligned, so normally the two peaks are around 10:30am and 14:00pm.

But looking at the detailed half-hour forecasts, the peak time in the sensor value does correlate to the overall peak generation on the graph.

![image](https://github.com/user-attachments/assets/8c59a2ad-709b-4052-9aa8-c51a54dd166c)

**Reply by autoSteve** [reply]
Looks fine to me. A bit weird, given the extreme split of times on each array.

It'd be fun to see the individual sites charted alongside, but the half-hourly detail for site breakdown is not there.

(Why do I feel a request coming to get the detail there as an option...?)

**Reply by gcoan** [reply]
> It'd be fun to see the individual sites charted alongside, but the half-hourly detail for site breakdown is not there.
> 
> (Why do I feel a request coming to get the detail there as an option...?)

Way ahead of you on that one @autoSteve 

I had thought about this a few times over the last couple of weeks and decided that whilst it may be fun, it wasn't something I really needed to see, would add lots more data for not a lot of insight so I haven't requested it
